test/distributed/org/apache/cassandra/distributed/impl/AbstractCluster.java

```diff
@@ -241,26 +249,53 @@ public void uncaughtException(Thread thread, Throwable throwable)
         }
     }
 
-    protected AbstractCluster(File root, Versions.Version initialVersion, List<IInstanceConfig> configs,
-                              ClassLoader sharedClassLoader)
+    protected AbstractCluster(AbstractBuilder builder)
     {
-        this.root = root;
-        this.sharedClassLoader = sharedClassLoader;
+        this.root = builder.getRoot();
+        this.sharedClassLoader = builder.getSharedClassLoader();
+        this.subnet = builder.getSubnet();
+        this.tokenSupplier = builder.getTokenSupplier();
+        this.nodeIdTopology = builder.getNodeIdTopology();
+        this.configUpdater = builder.getConfigUpdater();
         this.instances = new ArrayList<>();
         this.instanceMap = new HashMap<>();
-        this.initialVersion = initialVersion;
-        int generation = AbstractCluster.generation.incrementAndGet();
+        this.initialVersion = builder.getVersion();
+        this.filters = new MessageFilters();
 
-        for (IInstanceConfig config : configs)
+        int generation = AbstractCluster.generation.incrementAndGet();
+        for (int i = 0; i < builder.getNodeCount(); ++i)
         {
+            int nodeNum = i + 1;
+            InstanceConfig config = createInstanceConfig(nodeNum);
+
             I instance = newInstanceWrapperInternal(generation, initialVersion, config);
             instances.add(instance);
             // we use the config().broadcastAddressAndPort() here because we have not initialised the Instance
             I prev = instanceMap.put(instance.broadcastAddress(), instance);
             if (null != prev)
                 throw new IllegalStateException("Cluster cannot have multiple nodes with same InetAddressAndPort: " + instance.broadcastAddress() + " vs " + prev.broadcastAddress());
         }
-        this.filters = new MessageFilters();
+    }
+
+    public InstanceConfig newInstanceConfig()
+    {
+        return createInstanceConfig(size() + 1);
+    }
+
+    private InstanceConfig createInstanceConfig(int nodeNum)
```

Comment:
Do you think it might be better to leave `createInstanceConfig` in `Builder` and use `IntFunction` as a factory method instead? I'm just wondering how much it's going to grow, really, and keep in mind that this is the code that we'll inevitably duplicate 4 times.

Reply:
two things: "leave createInstanceConfig in Builder" and "we'll inevitably duplicate 4 times."

"leave createInstanceConfig in Builder"

I think the builder should be simple and small else it can get in the way.  It also gets ambiguous to semantics as in the case of the bootstrap test.  The cluster was created with one set of configs, the node is created with a different setup (maybe for a different cluster); this means its easy to use the API incorrectly.  It is simpler to reason about if the builder is just that, a set of methods to aid in calling `new` on the cluster class.

"we'll inevitably duplicate 4 times."

May make sense to have a utility function that provides this, you can choose to use it or not but its there to replace the need to write multiple times. And if you need to fork, nothing is getting in your way.

So I am fine with a utility method that has the logic, but don't think it should live in the Builder

Reply:
I tried to extract it out and I am changing my mind on common util.  The issue is that the actual logic to create is `org.apache.cassandra.distributed.impl.InstanceConfig#generate` which lives in implantation and not the API, so to have a factory with all the same methods leaks those implementation details.  

I can add the following utils though

```
public static String getAddress(AbstractBuilder<?, ?, ?> builder, int nodeNum)
    {
        return "127.0." + builder.getSubnet() + "." + nodeNum;
    }
    
    public static NetworkTopology getNetworkTopology(AbstractBuilder<?, ?, ?> builder)
    {
        String ipPrefix = "127.0." + builder.getSubnet() + ".";
        return NetworkTopology.build(ipPrefix, builder.getBroadcastPort(), builder.getNodeIdTopology());
    }
```

This assumes you do the local network, but doesn't force that on you if you choose not too.  With that, the below would live in implementation

```
String ipAddress = getAddress(builder, nodeNum);

        // for impl
        String seedIp = getSubnetPrefix(builder) + ".1";
        long token = builder.getTokenSupplier().token(nodeNum);

        NetworkTopology topology = getNetworkTopology(builder);

        InstanceConfig config = InstanceConfig.generate(nodeNum, ipAddress, topology, root, String.valueOf(token), seedIp);
        if (configUpdater != null)
            configUpdater.accept(config);

        return config;
```

Reply:
even went down that road and given the usage in abstract cluster, relying on the builder doesn't help...  I could create a interface but that feels like overkill.  I think the simplest thing is to just duplicate.  If we really look at it, its really best to duplicate; here is why.

```
String ipPrefix = "127.0." + subnet + ".";
String seedIp = ipPrefix + "1";
```

This is implementation specific.  If a cluster would rather the seed be the last node, why not?  So seed can't be in the core interface

```
long token = tokenSupplier.token(nodeNum);
```

What value does the builder provide since the token supplier does all the work?

```
NetworkTopology topology = NetworkTopology.build(ipPrefix, broadcastPort, nodeIdTopology);
```

ip prefix is implementation specific, and not sure why we can't just call it.  what if we wanted a different implementation or method call?

```
InstanceConfig config = InstanceConfig.generate(nodeNum, ipAddress, topology, root, String.valueOf(token), seedIp);
```

This is already defined in the implementation so not a issue

```
if (configUpdater != null)
            configUpdater.accept(config);
```

I guess would be nicer to have a no-op implementation rather than nullable.

given all this, I don't really see the value of a utility method as it either leaks implementation details or adds more complication.

Reply:
I was thinking about specific code that we have right now, not about theoretical cluster implementation that would change seed, tbh. I'm mostly concerned about us having to modify this code in several places instead of one in the future, but since the goal of API was to create a common interface and not to reduce the amount of duplication, I don't mind.

Reply:
Should we leave `getSubnetPrefix` in API? If yes, maybe we should rename it to `default` or something like that. I don't have a strong opinion about that though.